### PR TITLE
SDIT-1398: 🎨 Switch to aws kotlin sdk

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt [hunk 1]
```diff
@@ -34,45 +32,49 @@ class S3Config(private val s3Properties: S3Properties) {
...
   @Bean
   @ConditionalOnProperty(name = ["s3.provider"], havingValue = "aws")
-  fun s3Client(): S3AsyncClient = with(s3Properties) {
-    S3AsyncClient.builder()
-      .credentialsProvider(DefaultCredentialsProvider.builder().build())
-      .region(Region.of(region))
-      .build().also {
-        log.info("Creating AWS S3Client with DefaultCredentialsProvider and region {}", region)
-      }
+  fun s3Client(): S3Client = runBlocking {
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt:35]: have to use `runBlocking` here when creating beans

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt [hunk 2]
```diff
@@ -34,45 +32,49 @@ class S3Config(private val s3Properties: S3Properties) {
...
-            log.info("Creating S3 bucket {} as it was not found", bucketName)
-            this.createBucket(CreateBucketRequest.builder().bucket(bucketName).build()).thenApply {
-              "Bucket $bucketName now created"
+        try {
+          this.headBucket { bucket = bucketName }
+          log.info("Bucket {} exists already", bucketName)
+        } catch (e: NotFound) {
+          this.createBucket {
+            bucket = bucketName
+            createBucketConfiguration {
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt:70]: seems a bit naff but the kotlin sdk requires this extra configuration to create the bucket

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt [hunk 3]
```diff
@@ -34,45 +32,49 @@ class S3Config(private val s3Properties: S3Properties) {
...
+    with(s3Properties) {
+      S3Client.fromEnvironment {
+        this.region = s3Properties.region
+        this.endpointUrl = Url.parse(localstackUrl!!)
+        this.forcePathStyle = true
+        this.credentialsProvider = StaticCredentialsProvider {
+          accessKeyId = "any"
+          secretAccessKey = "any"
+        }
+        this.useArnRegion = true
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt:55]: Might be nicer not to use `with(s3Properties)` and drop all the `this` here? I get confused by overlapping this's.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt (hunk 1 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt:35]: I guess there is no concept of a bean returning a suspend function or anything? Not that the blocking matters in this case

**Review by andymarke — APPROVED** [review]
lgtm

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt (hunk 3 above)
Comment (petergphillips) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerdownloadapi/config/S3Config.kt:55]: agreed, will fix in next PR

**Pull request merged**